### PR TITLE
MXNet frontend support for AMP cast op

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

Support `amp_cast` and `amp_multicast` used by mxnet mixed precision. @icemelon9

### CODE REVIEW

**Comment by junrushao1994** [comment]
Hmmm just to confirm, do we really want to keep those "fake" operators in the NNVM IR?

File: python/tvm/relay/frontend/mxnet.py [hunk 1]
```diff
@@ -903,6 +903,22 @@ def _mx_resize(inputs, attrs):
...
+    dtypes = [_infer_type(x).checked_type.dtype for x in inputs]
+    supported_dtypes = ['float16', 'float32']
+    assert all([x in supported_dtypes for x in dtypes]), \
+            "amp_multicast support is limited to float16 and float32 inputs only."
+    dtype = 'float32' if cast_narrow else dtypes[0]
+    for t in dtypes:
+        if cast_narrow and t == 'float16':
+            dtype = 'float16'
+            break
+        elif not cast_narrow and t == 'float32':
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py:917]:
```suggestion
        if not cast_narrow and t == 'float32':
```

File: python/tvm/relay/frontend/mxnet.py [hunk 2]
```diff
@@ -903,6 +903,22 @@ def _mx_resize(inputs, attrs):
     return _op.image.resize(inputs[0], size,
                             coordinate_transformation_mode="align_corners")
 
+def _mx_amp_multicast(inputs, attrs):
+    cast_narrow = attrs.get_bool("cast_narrow", False)
+    dtypes = [_infer_type(x).checked_type.dtype for x in inputs]
+    supported_dtypes = ['float16', 'float32']
+    assert all([x in supported_dtypes for x in dtypes]), \
+            "amp_multicast support is limited to float16 and float32 inputs only."
+    dtype = 'float32' if cast_narrow else dtypes[0]
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py:912]:
```suggestion
    dtype = dtypes[0]
```

File: python/tvm/relay/frontend/mxnet.py [hunk 3]
```diff
@@ -903,6 +903,22 @@ def _mx_resize(inputs, attrs):
...
+    for t in dtypes:
+        if cast_narrow and t == 'float16':
+            dtype = 'float16'
+            break
+        elif not cast_narrow and t == 'float32':
+            dtype = 'float32'
+            break
+    return [relay.cast(x, dtype) for x in inputs]
```
Comment (icemelon9) [inline python/tvm/relay/frontend/mxnet.py:920]:
```suggestion
    return [_op.cast(x, dtype) for x in inputs]
```

File: tests/python/frontend/mxnet/test_forward.py [hunk 4]
```diff
@@ -1131,6 +1131,52 @@ def verify(a_np, b_np):
     verify(np.asarray([1.0], 'float32'), np.asarray([2.0],'float32'))
     verify(np.asarray([4.0], 'float32'), np.asarray([3.0],'float32'))
 
+def test_forward_amp_cast():
+    def verify(from_dtype, to_dtype):
+        from_nd = mx.nd.ones((2,2), dtype=from_dtype)
```
Comment (icemelon9) [inline tests/python/frontend/mxnet/test_forward.py:1136]: Could you use random tensors with larger size?

File: tests/python/frontend/mxnet/test_forward.py [hunk 5]
```diff
@@ -1131,6 +1131,52 @@ def verify(a_np, b_np):
...
+                op_res = intrp.evaluate()(from_np)
+                assert op_res.dtype == to_dtype, op_res.dtype
+                tvm.testing.assert_allclose(op_res.asnumpy(), 1.)
+
+    verify('float32', 'float16')
+    verify('float16', 'float32')
+
+def test_forward_amp_multicast():
+    def verify(dtypes, cast_narrow, expected_dtype):
+        x_nps = [np.ones((2,2), dtype=dtype) for dtype in dtypes]
```
Comment (icemelon9) [inline tests/python/frontend/mxnet/test_forward.py:1155]: same here

File: tests/python/frontend/mxnet/test_forward.py [hunk 6]
```diff
@@ -1307,6 +1353,8 @@ def verify(batch, seq_length, num_heads, head_dim):
 
 
 if __name__ == '__main__':
+    test_forward_amp_multicast()
```
Comment (icemelon9) [inline tests/python/frontend/mxnet/test_forward.py:1356]: let's move to the end of test

File: python/tvm/relay/frontend/mxnet.py (hunk 2 above)
Comment (anijain2305) [reply python/tvm/relay/frontend/mxnet.py:912]: Just a minor and soft suggestion, so feel free to ignore.  Maybe following has better readability

~~~
has_float16 = any(x == "float16" for x in dtypes)
has_float32 = any(x == "float32" for x in dtypes)

# logic with three vars - cart_narrow, has_float16, has_float32
~~~

**Comment by icemelon9** [comment]
Thanks @eric-haibin-lin @junrushao1994 @anijain2305

**Pull request merged**